## Numpy assignment

## 1 Preliminaries

In [52]:
# The usual preamble
import numpy as np

## Numpy

In [53]:
import numpy as np
x = np.array([2,3,1,0])
y =np.arange(10)
z = 2 + x

Write expressions in the next cell to retrieve 1 from `x`, 3 from `y`, and 3 from `z`.

In [54]:
print(x[2]) #prints the 3rd element of 1D array x
print(y[3]) #prints the 4th element of 1D array y
print(z[2]) #prints the 3rd element of 1D array z

1
3
3


In the next cell, write an expression that generates a 5 by 5 array filled with zeros, and another  that generates a 1 by 1 array with a 1 in it.

In [55]:
X = np.zeros( (5,5) ) # np.zeros generates 2D array of 5x5 filled with zero
X

array([[0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.]])

In [56]:
Y = np.ones( (1,1) ) # np.zeros generates 2D array of 1x1 filled with ones
Y 

array([[1.]])

<style>
    @font-face {
        font-family: "Computer Modern";
        src: url('http://mirrors.ctan.org/fonts/cm-unicode/fonts/otf/cmunss.otf');
    }
    div.cell{
        width:800px;
        margin-left:16% !important;
        margin-right:auto;
    }
    h1 {
        font-family: Helvetica, serif;
    }
    h4{
        margin-top:12px;
        margin-bottom: 3px;
       }
    div.text_cell_render{
        font-family: Computer Modern, "Helvetica Neue", Arial, Helvetica, Geneva, sans-serif;
        line-height: 145%;
        font-size: 130%;
        width:800px;
        margin-left:auto;
        margin-right:auto;
    }
    .CodeMirror{
            font-family: "Source Code Pro", source-code-pro,Consolas, monospace;
    }
    .text_cell_render h5 {
        font-weight: 300;
        font-size: 22pt;
        color: #4057A1;
        font-style: italic;
        margin-bottom: .5em;
        margin-top: 0.5em;
        display: block;
    }
    
    .warning{
        color: rgb( 240, 20, 20 )
        }  

In the next cell, write an expression that uses an assignment to a splice to make all the negative values in `a` be 0. 

In [57]:
a = np.arange(10,-1,-1) - 3
print(a)
a[8:] = 0 # It will make all the elements zero for which value is negative
print(a)

[ 7  6  5  4  3  2  1  0 -1 -2 -3]
[7 6 5 4 3 2 1 0 0 0 0]


One way to do this is:

In [58]:
a[-3:] = 0
a

array([7, 6, 5, 4, 3, 2, 1, 0, 0, 0, 0])

But see if you can do it by writing an expression that make all negative values in `a`
be 0 no matter where they occur in `a`.  Hint: Use a Boolean conditipn

In [59]:
a = np.arange(10,-1,-1) - 3
print(a)
a = np.where(a < 0, 0, a)
print(a)

[ 7  6  5  4  3  2  1  0 -1 -2 -3]
[7 6 5 4 3 2 1 0 0 0 0]


Another option is to use `numpy` `where`, combined with an assignment.
The form is

```
>>> np.where(condition, x, y)
```

where `condition`, `x` and `y` are all arrays of the same size (or are broadcastable to the same size).
Return an array containing the members of `x` for which `condition` is `True`; otherwise return the corresponding member  of `y`.  It can be read "`x` where `condition`, else
`y`".

```
>>> a = np.array([1, 4.1, 9, 16])
>>> b = np.array([1,2,3,4])
>>> np.where((np.sqrt(a) - b) == 0,np.sqrt(a),b)
array([1., 2., 3., 4.])
```

In [60]:
a = np.arange(10,-1,-1) - 3
 # it uses the where clause to replace negative elements with 0 in an array
a = np.where(a < 0, 0, a)

In [61]:
a

array([7, 6, 5, 4, 3, 2, 1, 0, 0, 0, 0])

In the next cell write an expression that produces an array containing result of multiplying 7 by each of the first 12 integers (1 - 12). There's a hard way to do this and an easy way.  The easy way uses elementwise operations.

In [62]:
7 * np.arange(1,13)

array([ 7, 14, 21, 28, 35, 42, 49, 56, 63, 70, 77, 84])

In [63]:
b = np.arange(1,40,2)
print(b) # Created an array of of 20 elements

# First 12 elements of b are multiplied by 7 with elementwise operation.
b[:12] * 7 

[ 1  3  5  7  9 11 13 15 17 19 21 23 25 27 29 31 33 35 37 39]


array([  7,  21,  35,  49,  63,  77,  91, 105, 119, 133, 147, 161])

## 2D arrays

In [64]:
import numpy as np
room_matrix = \
np.array(
[[6,  3, 4, 1],
[5,  2, 3, 2],
[8,  3, 6, 2],
[5,  1, 3, 1],
[10, 4, 7, 2]])

cost_vector = np.array([40, 175, 90, 450])

In the next cell write an express that retrieves the last column of `room_matrix`.

In [65]:
# prints the values of last colomn of room_matrix array
room_matrix[:,3] 

array([1, 2, 2, 1, 2])

In the next cell, write an expression that retrieves the third row of `room_matrix`.

In [66]:
# This will return 3rd row (with index 2) of room_matrix array
room_matrix[2,:] 

array([8, 3, 6, 2])

In the next cell write an expression that retrieves the first two elements of the second column of `room_matrix`

In [67]:
# This will return 1st two elements of 3rd row (with index 2) of room_matrix array
room_matrix[:2,1] 

array([3, 2])

In the next cell write an expression that retrieves the following submatrix from `room_matrix`:

```
array([[2,3],
       [3,6]])
```

In [68]:
# This will return 2nd and 3rd row elements of 2nd and 3rd colomns
room_matrix[1:3,1:3]

array([[2, 3],
       [3, 6]])

In the next cell, write an expression that returns a Boolean array that identifies the members of `room_matrix` which are greater than or equal to 7.

In [69]:
# This returns a boolean array of room_matrix which identifies the elements greater than or equal to 7
print(room_matrix >= 7) 

[[False False False False]
 [False False False False]
 [ True False False False]
 [False False False False]
 [ True False  True False]]


In the next cell, write an expression that produces an array containing the items in `room_matrix` which are greater than or equal to 7.  Note:  This question and the previous question do not have the same answer.

In [70]:
# This returns an array of elements of room_matrix for which the value greater than or equal to 7
print(room_matrix[room_matrix >= 7])

[ 8 10  7]


## Iris data exercise

In the next cell, use a Boolean mask to return all the rows of the iris data set whose column 3 value is greater than 1.8 (I mean the column whose **index** is 3, the last column).
The cell is not empty because you have been provided with the code which loads the data.
So you should keep that code and **add to it**.

In [71]:
from sklearn.datasets import load_iris
data = load_iris()
features = data['data']
target = data['target']

In [72]:
# Returns all the rows of iris data whose last column value is greater than 1.8. It uses the boolean mask.
print((features[features[:,3] > 1.8]).shape)
features[features[:,3] > 1.8]

(34, 4)


array([[6.3, 3.3, 6. , 2.5],
       [5.8, 2.7, 5.1, 1.9],
       [7.1, 3. , 5.9, 2.1],
       [6.5, 3. , 5.8, 2.2],
       [7.6, 3. , 6.6, 2.1],
       [7.2, 3.6, 6.1, 2.5],
       [6.5, 3.2, 5.1, 2. ],
       [6.4, 2.7, 5.3, 1.9],
       [6.8, 3. , 5.5, 2.1],
       [5.7, 2.5, 5. , 2. ],
       [5.8, 2.8, 5.1, 2.4],
       [6.4, 3.2, 5.3, 2.3],
       [7.7, 3.8, 6.7, 2.2],
       [7.7, 2.6, 6.9, 2.3],
       [6.9, 3.2, 5.7, 2.3],
       [5.6, 2.8, 4.9, 2. ],
       [7.7, 2.8, 6.7, 2. ],
       [6.7, 3.3, 5.7, 2.1],
       [6.4, 2.8, 5.6, 2.1],
       [7.4, 2.8, 6.1, 1.9],
       [7.9, 3.8, 6.4, 2. ],
       [6.4, 2.8, 5.6, 2.2],
       [7.7, 3. , 6.1, 2.3],
       [6.3, 3.4, 5.6, 2.4],
       [6.9, 3.1, 5.4, 2.1],
       [6.7, 3.1, 5.6, 2.4],
       [6.9, 3.1, 5.1, 2.3],
       [5.8, 2.7, 5.1, 1.9],
       [6.8, 3.2, 5.9, 2.3],
       [6.7, 3.3, 5.7, 2.5],
       [6.7, 3. , 5.2, 2.3],
       [6.3, 2.5, 5. , 1.9],
       [6.5, 3. , 5.2, 2. ],
       [6.2, 3.4, 5.4, 2.3]])

In the next cell, use a Boolean array to return all the rows representing members of class 2 in the iris data set (the Boolean array is not the answer; it is used in retrieving the answer).

In [73]:
# Returns all the rows of iris data whose for which the class is 2.
print((features[target == 2]).shape) # prints the shape of array
features[target == 2]

(50, 4)


array([[6.3, 3.3, 6. , 2.5],
       [5.8, 2.7, 5.1, 1.9],
       [7.1, 3. , 5.9, 2.1],
       [6.3, 2.9, 5.6, 1.8],
       [6.5, 3. , 5.8, 2.2],
       [7.6, 3. , 6.6, 2.1],
       [4.9, 2.5, 4.5, 1.7],
       [7.3, 2.9, 6.3, 1.8],
       [6.7, 2.5, 5.8, 1.8],
       [7.2, 3.6, 6.1, 2.5],
       [6.5, 3.2, 5.1, 2. ],
       [6.4, 2.7, 5.3, 1.9],
       [6.8, 3. , 5.5, 2.1],
       [5.7, 2.5, 5. , 2. ],
       [5.8, 2.8, 5.1, 2.4],
       [6.4, 3.2, 5.3, 2.3],
       [6.5, 3. , 5.5, 1.8],
       [7.7, 3.8, 6.7, 2.2],
       [7.7, 2.6, 6.9, 2.3],
       [6. , 2.2, 5. , 1.5],
       [6.9, 3.2, 5.7, 2.3],
       [5.6, 2.8, 4.9, 2. ],
       [7.7, 2.8, 6.7, 2. ],
       [6.3, 2.7, 4.9, 1.8],
       [6.7, 3.3, 5.7, 2.1],
       [7.2, 3.2, 6. , 1.8],
       [6.2, 2.8, 4.8, 1.8],
       [6.1, 3. , 4.9, 1.8],
       [6.4, 2.8, 5.6, 2.1],
       [7.2, 3. , 5.8, 1.6],
       [7.4, 2.8, 6.1, 1.9],
       [7.9, 3.8, 6.4, 2. ],
       [6.4, 2.8, 5.6, 2.2],
       [6.3, 2.8, 5.1, 1.5],
       [6.1, 2

Extra credit:  In the next cell write an expression -- or expressions -- which returns
all the members of class 2 in the data whose column 3 value is greater than 1.8.

Hint: Set a variable to the result of one of your answers to the last two questions.
That result is an array, and you can do more indexing on it.

Two Solutions.  The one not hinted at uses Boolean array conjunction.

Solution 2.  This one uses `&`.  Say you have X an array of numbers and you want a Boolean array
that tells you exactly which numbers are even (`X%2 == 0`) and greater than 4 (`X > 4`).  You can
combine these two conditions using `7`:

In [74]:
X = np.arange(15).reshape((3,5))
print(X)
(X%2 == 0) & (X > 4)

[[ 0  1  2  3  4]
 [ 5  6  7  8  9]
 [10 11 12 13 14]]


array([[False, False, False, False, False],
       [False,  True, False,  True, False],
       [ True, False,  True, False,  True]])

To find solution 2, write a single expression that uses `&` which returns
all the rows of the Iris data class 2 in the data whose column 3 value is greater than 1.8.

This should produce a 2D array with 4 columns and 34 rows.

In [75]:
# Returns all the rows of iris data whose for which belongs to class is 2 and 3rd column value is greater than 1.8.
iris_col_3_class_2 = features[(features[:,3] > 1.8) & (target == 2)]
print(iris_col_3_class_2.shape)
print(iris_col_3_class_2)

(34, 4)
[[6.3 3.3 6.  2.5]
 [5.8 2.7 5.1 1.9]
 [7.1 3.  5.9 2.1]
 [6.5 3.  5.8 2.2]
 [7.6 3.  6.6 2.1]
 [7.2 3.6 6.1 2.5]
 [6.5 3.2 5.1 2. ]
 [6.4 2.7 5.3 1.9]
 [6.8 3.  5.5 2.1]
 [5.7 2.5 5.  2. ]
 [5.8 2.8 5.1 2.4]
 [6.4 3.2 5.3 2.3]
 [7.7 3.8 6.7 2.2]
 [7.7 2.6 6.9 2.3]
 [6.9 3.2 5.7 2.3]
 [5.6 2.8 4.9 2. ]
 [7.7 2.8 6.7 2. ]
 [6.7 3.3 5.7 2.1]
 [6.4 2.8 5.6 2.1]
 [7.4 2.8 6.1 1.9]
 [7.9 3.8 6.4 2. ]
 [6.4 2.8 5.6 2.2]
 [7.7 3.  6.1 2.3]
 [6.3 3.4 5.6 2.4]
 [6.9 3.1 5.4 2.1]
 [6.7 3.1 5.6 2.4]
 [6.9 3.1 5.1 2.3]
 [5.8 2.7 5.1 1.9]
 [6.8 3.2 5.9 2.3]
 [6.7 3.3 5.7 2.5]
 [6.7 3.  5.2 2.3]
 [6.3 2.5 5.  1.9]
 [6.5 3.  5.2 2. ]
 [6.2 3.4 5.4 2.3]]


## Boston Housing data exercise (UCI ML housing dataset)

In [76]:
from sklearn import datasets
boston = datasets.load_boston()
X,Y = boston.data,boston.target
print(X.shape, Y.shape)

(506, 13) (506,)


/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function load_boston is deprecated; `load_boston` is deprecated in 1.0 and will be removed in 1.2.

    The Boston housing prices dataset has an ethical problem. You can refer to
    the documentation of this function for further details.

    The scikit-learn maintainers therefore strongly discourage the use of this
    dataset unless the purpose of the code is to study and educate about
    ethical issues in data science and machine learning.

    In this special case, you can fetch the dataset from the original
    source::

        import pandas as pd
        import numpy as np


        data_url = "http://lib.stat.cmu.edu/datasets/boston"
        raw_df = pd.read_csv(data_url, sep="\s+", skiprows=22, header=None)
        data = np.hstack([raw_df.values[::2, :], raw_df.values[1::2, :2]])
        target = raw_df.values[1::2, 2]

    Alternative datasets include the California housing dataset (i.e.

The cell above loads the `sklearn` Boston housing data, which contains information about housing prices
for 501 houses sold in Boston, and the values of various attributes which have been found to be useful
in predicting home prices:

```
Index
0         - CRIM     per capita crime rate by town
1         - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.
2         - INDUS    proportion of non-retail business acres per town
3         - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
4         - NOX      nitric oxides concentration (parts per 10 million)
5         - RM       average number of rooms per dwelling
6         - AGE      proportion of owner-occupied units built prior to 1940
7         - DIS      weighted distances to five Boston employment centres
8         - RAD      index of accessibility to radial highways
9         - TAX      full-value property-tax rate per $10,000
10        - PTRATIO  pupil-teacher ratio by town
11        - B        1000(Bk - 0.63)^2 where Bk is the proportion of blacks by town
12        - LSTAT    % lower status of the population
```

The array `X`, defined in the cell above, is a 501x13 array containing the values of all
these variables for 501 Boston houses.  

The information stored in Y, the value to be predicted is

```
       - MEDV     Median value of owner-occupied homes in $1000's
 
```

1.  Construct an expression or expressions that creates a new table that omits all houses whose tracts
    bound the Charles River.
2.  Construct an expression or expressions that creates a new table containing only the CRIM column and
    the ZN column.  This is  our new X.  Also create a new Y containing the pupil-teach ratio.
    This is to help someone research how to predict poor education conditions.  The shape of the new X should
    (506,2) and the shape of the new Y should be (506,).  Show that you checked this in your answer.

In [77]:
# The new table returns all the rows of Boston housing data whose tracts do not bound the Charles River.
new_table = X[X[:,3] == 0]
print(new_table.shape)
print(new_table)

(471, 13)
[[6.3200e-03 1.8000e+01 2.3100e+00 ... 1.5300e+01 3.9690e+02 4.9800e+00]
 [2.7310e-02 0.0000e+00 7.0700e+00 ... 1.7800e+01 3.9690e+02 9.1400e+00]
 [2.7290e-02 0.0000e+00 7.0700e+00 ... 1.7800e+01 3.9283e+02 4.0300e+00]
 ...
 [6.0760e-02 0.0000e+00 1.1930e+01 ... 2.1000e+01 3.9690e+02 5.6400e+00]
 [1.0959e-01 0.0000e+00 1.1930e+01 ... 2.1000e+01 3.9345e+02 6.4800e+00]
 [4.7410e-02 0.0000e+00 1.1930e+01 ... 2.1000e+01 3.9690e+02 7.8800e+00]]


In [78]:
# The X_new table returns all the rows of Boston housing data for only CRIM column and the ZN column.
X_new = X[:,0:2]
print(X_new)
print()
print(X_new.shape)
print()
# The Y_new table returns all the rows of Boston target housing data which contains pupil-teach ratio
Y_new = Y[X[:,10] != 0]
print(Y_new)
print()
print(Y_new.shape)

[[6.3200e-03 1.8000e+01]
 [2.7310e-02 0.0000e+00]
 [2.7290e-02 0.0000e+00]
 ...
 [6.0760e-02 0.0000e+00]
 [1.0959e-01 0.0000e+00]
 [4.7410e-02 0.0000e+00]]

(506, 2)

[24.  21.6 34.7 33.4 36.2 28.7 22.9 27.1 16.5 18.9 15.  18.9 21.7 20.4
 18.2 19.9 23.1 17.5 20.2 18.2 13.6 19.6 15.2 14.5 15.6 13.9 16.6 14.8
 18.4 21.  12.7 14.5 13.2 13.1 13.5 18.9 20.  21.  24.7 30.8 34.9 26.6
 25.3 24.7 21.2 19.3 20.  16.6 14.4 19.4 19.7 20.5 25.  23.4 18.9 35.4
 24.7 31.6 23.3 19.6 18.7 16.  22.2 25.  33.  23.5 19.4 22.  17.4 20.9
 24.2 21.7 22.8 23.4 24.1 21.4 20.  20.8 21.2 20.3 28.  23.9 24.8 22.9
 23.9 26.6 22.5 22.2 23.6 28.7 22.6 22.  22.9 25.  20.6 28.4 21.4 38.7
 43.8 33.2 27.5 26.5 18.6 19.3 20.1 19.5 19.5 20.4 19.8 19.4 21.7 22.8
 18.8 18.7 18.5 18.3 21.2 19.2 20.4 19.3 22.  20.3 20.5 17.3 18.8 21.4
 15.7 16.2 18.  14.3 19.2 19.6 23.  18.4 15.6 18.1 17.4 17.1 13.3 17.8
 14.  14.4 13.4 15.6 11.8 13.8 15.6 14.6 17.8 15.4 21.5 19.6 15.3 19.4
 17.  15.6 13.1 41.3 24.3 23.3 27.  50.  50.  50.  2

Write an exporession that computes how many housing tracts in the data set bound
the Charles River.

In [79]:
# This returns number of housing tracts bound the Charles River.
X_Char_River = X[X[:,3] != 0]
print("Number of housing tracts bound the Charles River: ", len(X_Char_River))

Number of housing tracts bound the Charles River:  35


Checks!

## Part two

The answers for the following section are available on the inclass Boolean exercies solution notebook.

# Boolean combinations

In this section we load forest covertype data.  This is a multi class dataset that has data for 7 different forest covertypes (stored in the `target` attribute).  There are 581,012 forest plots with 54 attributes each (stored ina 581012x54 array) The first ten attributes are numerical, the last 44 are Boolean (true/false) attributes. Each of the Boolean attributes represents a qualitative soil type attribute which is either present or absent.  We will refer to all the attributes by their column index.  Thus the first attribute is attribute 0 and the last (a Boolean attribute) is attribute 53.

In [80]:
#from sklearn.datasets import load_wine
#wdata = load_wine()
from sklearn.datasets import fetch_covtype
data = fetch_covtype()
print(data.data.shape)  # data.data is the 581012 x 54 array
print(data.target.shape)     # data.target contains the class for each instance

(581012, 54)
(581012,)


Th first 26 columns of the plot index 50_210:

In [81]:
data.data[50_210,:26]

array([2.988e+03, 2.250e+02, 2.000e+00, 5.520e+02, 9.200e+01, 5.492e+03,
       2.160e+02, 2.410e+02, 1.620e+02, 4.290e+03, 1.000e+00, 0.000e+00,
       0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00,
       0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00,
       0.000e+00, 0.000e+00])

The last 28 column of the plot index 50_210:

In [82]:
data.data[50_210,26:]

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

[The Boolean arrays and masks notebook](https://github.com/gawron/python-for-social-science/blob/master/numpy/02_06_Boolean_Arrays_and_Masks.ipynb) discusses combining Boolean arrays with Boolean operators `&` (conceptually 'and') and `|` (conceptually 'or').   We can add to that `~` (conceptually 'not') to complete the list.  Look at the examples there.

 
The `sklearn` description of the covertype dataset is printed in the cell below. For a fuller understanding of the attributes in the dataset you might want to look at [the original UCI data set description.](https://archive.ics.uci.edu/ml/datasets/Covertype)

Forest covertypes
---------------------

The samples in this dataset correspond to 30×30m patches of forest in the US,
collected for the task of predicting each patch's cover type,
i.e. the dominant species of tree.
There are seven covertypes, making this a multiclass classification problem.
Each sample has 54 features, described on the
[dataset's homepage](https://archive.ics.uci.edu/ml/datasets/Covertype>).
Some of the features are boolean indicators
while others are discrete or continuous measurements.

**Data Set Characteristics:**

    =================   ============
    Classes                        7
    Samples total             581012
    Dimensionality                54
    Features                     int
    =================   ============

The function `sklearn.datasets.fetch_covtype` will load the covertype dataset;
it returns a dictionary-like object
with the feature matrix in the ``data`` member
and the target values in ``target``.
The dataset will be downloaded from the web if necessary.

`data.data` attribute is a `numpy` array containing the array.  As an example to help with this problem.  To compute an an array containing just the rows in which attribute 0 is greater than 50 you do. 

In [83]:
WW = data.data > 50
print(WW.shape)

(581012, 54)


The covertypes (or dominant tree) for each forest plot are in `data.target`, which is often the attribute used to store the classes in an `sklearn` clasification dataset.  There is data for 581,012 plots
so this is a 1D array of length 581,012.  

There are 7 distinct fiorest tyoes that a plot can have.

In [84]:
print(len(data.target))
print(set(data.target))

581012
{1, 2, 3, 4, 5, 6, 7}


### The problem

1. Write an expression that retrieves all rows that a value greater than 300 in the column indexed 1 (second column) and a value equal to 1 for the column indexed 12. 

2. Write an expression that retrieves the rows from either class 3 or class 5  that have a value greater than 300 in the column indexed 1 (second column) and a value equal to 1 for the column indexed 12. 

3. Write an expression that counts the rows from either class 3 or class 5  that have a value greater than 300 in the column indexed 1 (second column) and a value equal to 1 for the column indexed 12. 

In [85]:
# This returns all rows that a value greater than 300 in the column indexed 1 (second column) 
# and a value equal to 1 for the column indexed 12.
# It uses the boolean test 
problem_1 = data.data[(data.data[:,1] > 300) & (data.data[:,12] == 1)]
print(problem_1.shape)
problem_1

(45076, 54)


array([[2780.,  346.,   13., ...,    0.,    0.,    0.],
       [2725.,  353.,   19., ...,    0.,    0.,    0.],
       [2747.,  329.,   19., ...,    0.,    0.,    0.],
       ...,
       [2619.,  336.,   13., ...,    0.,    0.,    0.],
       [2630.,  317.,   12., ...,    0.,    0.,    0.],
       [2633.,  309.,    9., ...,    0.,    0.,    0.]])

In [86]:
# This returns all rows from either class 3 or class 5 that have a value greater than 300 in the column indexed 1 
# and a value equal to 1 for the column indexed 12.
# It uses the boolea test
problem_2 = data.data[((data.target == 3) | (data.target == 5)) & ((data.data[:,1] > 300) & (data.data[:,12] == 1))]
problem_2

array([[2556.,  334.,   21., ...,    0.,    0.,    0.],
       [2566.,  326.,   11., ...,    0.,    0.,    0.],
       [2577.,  323.,   10., ...,    0.,    0.,    0.],
       ...,
       [2614.,  333.,    6., ...,    0.,    0.,    0.],
       [2621.,  359.,   10., ...,    0.,    0.,    0.],
       [2626.,  328.,   11., ...,    0.,    0.,    0.]])

In [87]:
# Counts the rows for probem 2
problem_3 = data.data[((data.target == 3) | (data.target == 5)) & ((data.data[:,1] > 300) & (data.data[:,12] == 1))]
print("Count: ", len(problem_3))
print(problem_3.shape)

Count:  1261
(1261, 54)
